### PR TITLE
feat(languages): add odin language

### LINKED ISSUE #2340: Add Odin language support
Odin is a very nice language and it would be nice if it got out of the box support.

https://odin-lang.org/

image

### PR DESCRIPTION
**Pull request by RobWalt**
This commit adds the odin language by using the grammar and highlighting
queries written by [MineBills repo](https://github.com/MineBill/tree-sitter-odin).

This commit resolves #2340 by just applying all the changes already
discussed in the issue.

### CODE REVIEW

**Comment by RobWalt** [comment]
+ I tested the functionality with the example also located in [MineBills repo](https://github.com/MineBill/tree-sitter-odin) and it worked on my machine perfectly fine.

**Comment by the-mikedavis** [comment]
If you run `cargo xtask docgen` and commit the results it'll clear up the :x: from the Docs CI

File: runtime/queries/odin/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+(function_declaration
+  name: (identifier) @function)
+
+(proc_group
+  (identifier) @function)
+
+; ; Identifiers
+
+(type_identifier) @type
+(field_identifier) @property
```
Comment (the-mikedavis) [inline runtime/queries/odin/highlights.scm:22]:
```suggestion
(field_identifier) @variable.other.member
```

File: runtime/queries/odin/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+; ; Literals
+
+[
+  (interpreted_string_literal)
+  (raw_string_literal)
+  (rune_literal)
+] @string
+
+(escape_sequence) @escape
```
Comment (the-mikedavis) [inline runtime/queries/odin/highlights.scm:125]:
```suggestion
(escape_sequence) @constant.character.escape
```

**Review by the-mikedavis — CHANGES_REQUESTED** [review]
a few suggested changes here to bring the captures in line with the ones helix uses: https://github.com/helix-editor/helix/blob/f1a77370cf679eb0f3a58cfb12deec8ff7fc6325/book/src/themes.md#syntax-highlighting

**Comment by RobWalt** [comment]
Hey, I just looked into your review. Sorry for adding the queries so carelessly. I think I understand the purpose of the file a bit better now. All of your suggested changes make sense.

File: runtime/queries/odin/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  (float_literal)
+  (imaginary_literal)
+] @number
+
+[
+  (true)
+  (false)
+  (nil)
+  (undefined)
+] @constant
```
[the remark this answers is not in the document]
Comment (RobWalt) [reply runtime/queries/odin/highlights.scm]: I changed `true` and `false` to be of type `@constant.builtin.boolean` and used your suggestion for the rest

**Comment by the-mikedavis** [comment]
No worries, queries typically need a bit of tuning anyways 😄

**Review by the-mikedavis — APPROVED** [review]
Looks good. Thanks!

**Comment by spindlebink** [comment]
Highlighting still isn't working for me, although it seems like Helix is recognizing `.odin` files correctly and Helix lists the Odin grammar as being both fetched and compiled. Also, I'm not aware of an explicit Odin styleguide, but the Odin project itself uses tabs for indentation, so the default indentation mode should probably be set accordingly.

**Comment by the-mikedavis** [comment]
> Highlighting still isn't working for me

Check out `hx --health odin`. If it shows that you don't have highlight queries then you'll need to link your runtime directory: https://github.com/helix-editor/helix/blob/7b287f636f72766356aadb95ba2f32636e0601d3/README.md?plain=1#L52-L55

> the Odin project itself uses tabs for indentation

Would you like to submit a PR to change it? If Odin itself uses tabs it probably makes sense to set tabs as the default (and then it can be overridden on a per-project basis with #1249).

**Pull request merged**